File: find_normal2.m
% 設定ここから

%% 画像設定 (make_input_model.pyの設定と同じにすること)
N_ROW  = 256;  % 画像の行（縦方向）の数
N_COL  = 256;  % 画像の列（横方向）の数
pic_num = 15;  % 作成した画像の数

%% 入力ディレクトリ (最後の / は不要)
INPUT_DIR = "input";

%% 出力先ディレクトリ (最後の / は不要)
OUTPUT_DIR = "output";

% S = importdata(strcat(INPUT_DIR,'/light_source.txt'));
%S = a;
img_pixel_value = zeros(N_ROW,N_COL,pic_num);

sn_es = zeros(N_ROW,N_COL,3);

%% 画像読み込み
for a = 1:pic_num
    LOAD_IMG = strcat(INPUT_DIR,'/save_gazo_hishatai',num2str(a),'.png');
    img_pixel_value(:,:,a) = imread(LOAD_IMG);
end

%% 法線推定
for i = 1:N_ROW
   for j = 1:N_COL
      intensity = reshape(img_pixel_value(i,j,:),[pic_num,1]);
      non_zero_index = find(intensity > 0); % 輝度が0じゃないindexについて
      
      intensity_n0 = intensity(non_zero_index);
      S_n0 = S(non_zero_index,:);
      if numel(intensity_n0) >= 3
          sn_tmp = pinv(S_n0) * intensity_n0;
      else
          sn_tmp = pinv(S) * intensity; % 有効な光源数なくてもとりあえず求める
      end
 
      if norm(sn_tmp) > 0
         sn_tmp = sn_tmp / norm(sn_tmp); 
      end
      sn_es(i,j,:) = sn_tmp;
   end
end

%% 誤差の評価
% 画像による確認 (定性的評価)
check_sn = (sn_es + 1) / 2;
imwrite(check_sn,strcat(OUTPUT_DIR,'/result2.ppm'));

% 数値による確認 (定量的評価)
error_sn = zeros(N_ROW,N_COL);
load(strcat(INPUT_DIR,'/sn_true.mat'));

count_pixel = 0;
sum_error = 0;

for i = 1:N_ROW
   for j = 1:N_COL
       sn_true_tmp = [sn_true(i,j,1) sn_true(i,j,2) sn_true(i,j,3)]';
       sn_es_tmp = [sn_es(i,j,1) sn_es(i,j,2) sn_es(i,j,3)]';
       
       if norm(sn_true_tmp) > 0
           error_rad = acos(dot(sn_true_tmp,sn_es_tmp));
           error_deg = rad2deg(error_rad);
           
           error_sn(i,j) = error_deg / 90;
           sum_error = sum_error + error_deg;
           count_pixel = count_pixel + 1;
       end
   end
end

ave_sn_error = sum_error / count_pixel
imwrite(error_sn,strcat(OUTPUT_DIR,'/error2.pgm'));

File: find_normal2.py
import math
import numpy as np
import cv2
import random

# 設定ここから

## 画像設定 (make_input_model.pyの設定と同じにすること)
N_ROW  = 128 # 画像の行 (縦方向) の数
N_COL  = 128 # 画像の列 (横方向) の数
pic_num =  7 # 作成した画像の数

## 入力ディレクトリ (最後の / は不要)
INPUT_DIR = "input"; 
## 出力先ディレクトリ (最後の / は不要)
OUTPUT_DIR = "output"; 


# 設定ここまで

## 各種初期化
sn_es = np.zeros([N_ROW, N_COL, 3])

LOAD_S_DIR = INPUT_DIR + "/light_source.txt"
S  = np.loadtxt(LOAD_S_DIR)
img_pixel_value = np.zeros([N_ROW,N_COL,pic_num])

## 画像読み込み
for a in range(0,pic_num):
    LOAD_IMG_DIR = INPUT_DIR + "/" +  str(a + 1) + ".pgm"
    img_pixel_value[:,:,a] = cv2.imread(LOAD_IMG_DIR, cv2.IMREAD_GRAYSCALE)

## 推定
for i in range(0,N_ROW):
    for j in range(0,N_COL):
        intensity_tmp = img_pixel_value[i,j,:] # np.array([img_pixel_value[i,j,:]]).T # 縦ベクトルを作成するため
        non_zero_index = np.where(intensity_tmp > 0)
        intensity_n0 = np.array([intensity_tmp[non_zero_index]]).T
        S_n0 = S[non_zero_index]
        if len(intensity_n0) >= 3:
            sn_tmp = np.dot(np.linalg.pinv(S_n0),intensity_n0) # 計算のときだけ縦ベクトルに
        else:
            intensity = np.array([img_pixel_value[i,j,:]]).T 
            sn_tmp = np.dot(np.linalg.pinv(S),intensity) # 有効な光源数がなくてもとりあえず求める
        if np.linalg.norm(sn_tmp) > 0:
                sn_tmp = sn_tmp / np.linalg.norm(sn_tmp)
                sn_es[i,j,:] = sn_tmp.T # 計算後は横ベクトルに戻す
     

## 誤差の評価
# 画像による確認 (定性的評価)
check_sn = (sn_es + 1) * 255 / 2
check_sn_bgr = check_sn[:, :, [2, 1, 0]] # CV2はBGRの順に認識されるため 
cv2.imwrite(OUTPUT_DIR + "/result2.ppm", check_sn_bgr) 

# 数値による確認 (定量的評価)
error_sn = np.zeros([N_ROW, N_COL])

LOAD_SNTRUE_DIR = INPUT_DIR + "/sn_true.npy"
sn_true = np.load(LOAD_SNTRUE_DIR)

count_pixel = 0
sum_error = 0

for i in range(0,N_ROW):
    for j in range(0,N_COL):
        sn_true_tmp = sn_true[i,j,:]
        sn_es_tmp = sn_es[i,j,:]
        if np.linalg.norm(sn_true_tmp) > 0:
            error_rad = np.arccos(np.dot(sn_es_tmp,sn_true_tmp))
            error_deg = math.degrees(error_rad)

            error_sn[i,j] = error_deg / 90 * 255
            sum_error += error_deg
            count_pixel += 1

ave_sn_error = sum_error / count_pixel
print(ave_sn_error)

cv2.imwrite(OUTPUT_DIR + "/error2.pgm", error_sn)